File: Binary_Tree/572_subtree_of_another/572.cpp
#include "../tree.cpp"
#include <algorithm>
#include <vector>


bool isSame(BinaryTree* p, BinaryTree* q) {
    if (p == nullptr && q == nullptr) {
        return true;
    }
    if ((p == nullptr || q == nullptr) || (p -> val != q -> val)) {
        return false;
    }
    return (isSame(p -> left, q -> left) && isSame(p -> right, q -> right));
}

bool isSubtree(BinaryTree* root, BinaryTree* subRoot) {
    if (root == nullptr) {
        return false;
    }
    bool compareTree = false;
    if (root -> val == subRoot -> val) {
        compareTree = isSame(root, subRoot);
    }
    return (isSubtree(root -> left, subRoot) || isSubtree(root -> right, subRoot) || compareTree);
}


int main() {
    std::vector<int> pArray = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
    std::vector<int> qArray = {5, 6, 7, 8, 9};
    
    BinaryTree* p = build_tree(pArray);
    print_tree(p);
    std::cout << " " << std::endl;

    BinaryTree* q = build_tree(qArray);
    print_tree(q);
    std::cout << " " << std::endl;
    
    std::cout << " is same : " << isSubtree(p, q) << std::endl;
}

File: Binary_Tree/572_subtree_of_another/572.rs
#[path = "../tree.rs"]
mod tree;


fn is_same(p: &Option<Box<tree::BinaryTree>>, q: &Option<Box<tree::BinaryTree>>) -> bool {

    match (p, q) {
        (None, None) => return true,
        (_, None) | (None, _) => return false,
        (Some(p), Some(q)) => {
            if p.val != q.val {
                return false;
            }
            return is_same(&p.left, &q.left) && is_same(&p.right, &q.right);
        } 
    }
}

fn is_sub_tree(root: &Option<Box<tree::BinaryTree>>, subRoot: &Option<Box<tree::BinaryTree>>) -> bool {

    let compare_tree = is_same(root, subRoot);

    match root {
        None => return false,
        Some(root) => {
            return is_sub_tree(&root.left, &subRoot) || is_sub_tree(&root.right, &subRoot) || compare_tree;
        }
    }
}




fn main() {
    let p = tree::build_tree(&mut vec!(2, 3, 4, 5, 6, 7, 8, 9, 10, 11));
    tree::print_tree(&p, 1);

    let q = tree::build_tree(&mut vec!(7, 9, 8, 10, 11));
    tree::print_tree(&q, 1);

    let result = is_sub_tree(&p, &q);
    println!("tree is same : {}", result);

}

